### ISSUE TITLE
Ability to lint template-generated code

### ISSUE THREAD

**Issue by v1adko** [plopjs/plop#80]
Hey everyone.

This is an awesome project, thank you! Stumbled upon it after playing around with [React Boilerplate](https://github.com/react-boilerplate/react-boilerplate) (another awesome project).

When I started writing more complex templates for React (for example conditional prop or nested conditions) in some "branches" styling would be really off (extra spaces, trailing commas, etc.). I know this could be avoided either by simplifying templates or writing multiple templates,  but I wanted to avoid both so I do it mostly by `if..else` (and it also becomes a bit harder if we `modify` existing code);

So to avoid generating templates with errors I have hacked around a way to run eslint (with `--fix`) as a last action if any of the other actions asked for it (you can check it out in [this snippet](https://gist.github.com/v1adko/441e8b7f8f8e45114d125de1ad8c67a3)).

Think it's quite useful and does not add too much overhead, so I wanted to share and maybe start a discussion. Ideally it would be a nice drop-in plugin but I have not figured out a good way to extend generators and I do it by extending the actions array on each generator initialization.

Additionally it can be configured to run any script of course (e.g prettier, css-lint, etc.)

Is there a better way to do it? Maybe by extending generators via a plugin?

**Comment by amwmedia** [comment]
Hey, this actually sounds like a perfect place for a custom action type! You can read more about how to create a custom action type under the documentation for [`setActionType`](https://plopjs.com/documentation/#setactiontype). Once you have your new action type setup you can actually distribute it as a `plop-pack` that can easily be loaded using [`plop.load()`](https://github.com/amwmedia/plop/blob/master/plop-load.md). I'm planning to do some screencasts soon around how to use all these different things, but if you need help in the mean time, I'm happy to show you how it all fits together.

Long story short, when all is said and done, someone else could consume your action by simply doing and `npm install` of the package, and including a single `plop.load("package-name")` in their plopfile. That would get them your action, available in their own plopfile.

**Comment by amwmedia** [comment]
I just realized that the `plop.load()` documentation doesn't mention `actions` but plop 1.8.0 introduced that as well, so that's the 4th thing that `plop.load()` can import.

**Comment by v1adko** [comment]
@amwmedia Thanks, looks great, I'll play with it.
Is it possible to reference other actions within an action? Linter needs an array of paths of generated/modified files and those are defined on other actions, so Ideally it would be cool to maybe have a `plop.getActions()` or something like that, just to avoid duplicating file paths.
Then within my custom action I could filter by a specific flag (like `shouldLintAfter`) and work from that. And that also would allow to set other certain conditions like "run prettier only on modified files", etc.

**Comment by amwmedia** [comment]
I don't believe that actions currently have data about the actions array that they are currently running in. However, the list of paths should be know by the person writing the generator. So it should be possible for your action to accept an array of path template strings to operate on. That would give the consumer complete power to have the linting happen on only what they'd like.

something like...

``` javascript
plop.setGenerator('something', {
  prompts: [...],
  actions: [{
    type: 'add',
    path: './{{name}}.js'
  }, {
    type: 'linter',
    paths: ['./{{name}}.js']
  }],
});
```

**Comment by v1adko** [comment]
Yes, that's exactly right, I just wanted to know if there's a way to avoid route duplication and run specific actions on the files affected by the generator. Something like that.

```
plop.setGenerator('something', {
  prompts: [...],
  plugins: [linter],
  actions: [{
    type: 'add',
    lint: true,
    path: './{{name}}.js'
  }],
});
```

And I'm currently achieving it by extending the actions array with a function.

**Comment by amwmedia** [comment]
Let me give this a little thought. If you even had the name of the generator that was currently running, you could use the `getGenerator()` method on the plop api to get the list of prompts and actions. Alternatively if you had access to the default actions via the `getActionType()` method you could create a custom `add` action that would delegate to the build-in action. However, currently `getActionType()` only has access to 3rd party action types.

**Comment by v1adko** [comment]
Yeah, no problem - that's all I wanted tbh, just some discussion. The API and functionality is great as is :)

**Comment by amwmedia** [comment]
well, the conversation has exposed a few shortcomings of the API that I didn't think about... so thanks! 👍  I still thing that a list of paths might be the best because it is very explicit and doesn't make any assumptions. Although I could see how having an option that you can mix into other actions could be very convenient. I'm going to leave this open as a reminder to implement something that would give the customAction a little context around the running generator.

**Comment by v1adko** [comment]
Awesome, thank you too! :)

**Comment by blling** [comment]
Hi, guys, how to prettier the generated js file?

**Comment by macrozone** [comment]
with my change described in here https://github.com/amwmedia/plop/issues/103

it might be possible to do something like:


```
plop.setActionType('add-with-prettier', (data, config, {runActions}) =>  {
  return runActions([
    {
      type: 'add',
      ...config
    },
    {
      type: 'prettier',
      path: config.path
    },
  ], data)
});
```

given you have a custom action named `prettier` or `lint`

**Comment by blling** [comment]
Where is your `prettier' action come from?

**Comment by macrozone** [comment]
nowhere. Its just an approach to solve the api discussion from above.

You would have to implement it as a custom action

**Comment by amwmedia** [comment]
There is also plop-prettier https://www.npmjs.com/package/plop-prettier

**Comment by macrozone** [comment]
@amwmedia yeah, but it duplicates the add-logic, which can be avoided by composing actions like in  https://github.com/amwmedia/plop/issues/103 :-)

**Comment by FDiskas** [comment]
I'm using Eslint v6.7.1 and the syntax is something like this
```
import { CLIEngine } from 'eslint';

// ...
    plop.setActionType('formatCode', function(answers, config, plopInstance) {
        if (config && config.path && plopInstance) {
            const filePath = plopInstance.renderString(config.path, answers);
            const eslint = new CLIEngine({ fix: true });
            const results = eslint.executeOnFiles([...filePath.split(',')]);

            CLIEngine.outputFixes(results);

            return 'code formatted';
        }

        throw new Error('Formatting skipped');
    });
```

And later I can use it like so:
```
{
     type: 'formatCode',
     path: 'app/screens/{{pascalCase name}}/{{pascalCase name}}Screen',
}
```

**Comment by crutchcorn** [comment]
Clsoing, since this is a question and there seems to be some community answers

**Comment by natterstefan** [comment]
Thanks, everyone for the collaboration and the proposed solution(s) here! I took the liberty of moving the code to a reusable package to avoid copy-pasting this again and again: [plop-action-eslint](https://github.com/natterstefan/plop-action-eslint).
